### PR TITLE
Added FlagRule

### PR DESCRIPTION
**Pull request by jglick**
A lot of Jenkins tests turn on some hidden flag defined by static variable, so it would be convenient to not need `@Before` & `@After` methods for this common idiom.

### CODE REVIEW

**Review by oleg-nenashev — APPROVED** [review]
LGTM. It might make sense to also add a convenience `SystemPropertyRule`, but it can be done later

**Comment by jglick** [comment]
Yeah I was thinking of adding a factory method for system properties specifically. OTOH _most_ places in Jenkins use a `static` field initialized from the system property so I am not sure it would get used very much.

I need to find a good demonstration usage of this API in something I own, unlike https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/405.

File: src/test/java/org/jvnet/hudson/test/FlagRuleNoReplacementTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        assertFalse(FLAG);
+    }
+
+    @AfterClass
+    public static void end() {
+        assertFalse(FLAG);
+    }
+
+    @Rule
+    public TestRule flagRule = new FlagRule<Boolean>(() -> FLAG, x -> FLAG = x);
```
Comment (jglick) [inline src/test/java/org/jvnet/hudson/test/FlagRuleNoReplacementTest.java:49]: As per [this](https://stackoverflow.com/a/27472025/12916) there is no slick syntax for representing field accessors. There _could_ be a factory like

```java
@Rule
public TestRule flagRule = FlagRule.field(ThisTest.class, "FLAG");
```

which would look up the field reflectively, but then this is opaque to IDEs looking to find usages etc., which does not seem to justify the slightly shorter syntax.

**Review by oleg-nenashev — APPROVED** [review]
Great stuff!

**Pull request merged**